### ISSUE TITLE
navigateTo raise 'nuxt instance unavailable' error after useFetch called

### ISSUE THREAD

**Issue by jrog612** [nuxt/nuxt#12342]
### Environment

- Operating System: `Linux`
- Node Version:     `v14.21.1`
- Nuxt Version:     `3.0.0`
- Nitro Version:    `1.0.0`
- Package Manager:  `yarn@1.22.10`
- Builder:          `vite`
- User Config:      `app`, `srcDir`, `modules`, `runtimeConfig`, `css`, `vite`
- Runtime Modules:  `@pinia/nuxt@0.4.6`
- Build Modules:    `-`

### Reproduction

https://stackblitz.com/edit/github-a99c1a?file=pages/after-fetch.vue

### Describe the bug

I am implementing social login such as Google login. Strangely, calling navigateTo after useFetch causes a problem. So I did a few tests and found out when this bug was occurring. 

**First**, access nuxt page without recommended ways like `<NuxtLink>` or `navigateTo`.

When you click `after-fetch` button, page moved by `location = link` code. And error was raised.
Then you click `after-fetch with recoomended ways` link, error was not raised.

**Second**, use nagivateTo with useFetch.

When you click `without-fetch` button, it works well. And also `only-fetch` too.

Most social logins use redirect. So I need to access my nuxt page from external site.

Thanks

**Comment by oritwoen** [comment]
I tested now this example in two version syntax `Options API` and `Sugar` and I can confirm that bug exist in `Options API` where on `Sugar` everything works fine.

Demo `Options` where `afterFetch` not worked: 
- https://stackblitz.com/edit/github-a99c1a?file=pages/after-fetch.vue

Demo `Sugar` where `afterFetch` worked fine:
- https://stackblitz.com/edit/github-a99c1a-i1fhmd?file=pages%2Fafter-fetch.vue,pages%2Fonly-fetch.vue

### So, this not work: 

```vue
<script>
export default {
  async setup() {
    const { data } = await useFetch('https://api.publicapis.org/random', {
      method: 'GET',
    });
    console.log(data.value);
    console.log('after-fetch: redirect!');
    return navigateTo(
      { path: '/', query: { from: 'after-fetch' } },
      { replace: true }
    );
  },
};
</script>

<template>
  <div>redirect to index</div>
</template>
```

### This work:
```vue
<script setup lang="ts">
const { data } = await useFetch('https://api.publicapis.org/random', {
  method: 'GET',
});

console.log(data.value);
console.log('after-fetch: redirect!');

await navigateTo(
  { path: '/', query: { from: 'after-fetch' } },
  { replace: true }
);
</script>

<template>
  <div>redirect to index</div>
</template>
```

FYI @danielroe

**Comment by danielroe** [comment]
This is Vue 3 behaviour; the Vue compiler transforms async script setup blocks to persist the component instance across an await. But this does not work with regular script blocks so you must use callWithNuxt if you want to use a Nuxt composable after an await.

This it certainly needs to be documented. We might also be able to add a feature to do an additional transform when using `defineNuxtComponent`.

Wdyt @antfu?

**Comment by Jamiewarb** [comment]
@danielroe was this documented at all, or do you have any examples for using `callWithNuxt` ?

**Comment by danielroe** [comment]
Yes, you can see https://nuxt.com/docs/guide/concepts/auto-imports#using-vue-and-nuxt-composables for some documentation.

**Comment by iamandrewluca** [comment]
Got the same error. A single line repositioning fixed the issue 😄 

<img width="1265" alt="image" src="https://user-images.githubusercontent.com/1881266/234620917-b2032250-ef55-44f5-9227-6019af4153d6.png">

**Comment by SebbeJohansson** [comment]
@danielroe found this issue while researching an issue I'm having from using an async composable (seemingly in a place too early). I noticed that the link you provided in march does not contain any `callWithNuxt` references. I would suggest that the documentation is updated with a link to something that references `runWithContext` instead of linking to https://github.com/nuxt/nuxt/issues/14269#issuecomment-1397352832

I really hope I didn't misunderstand something!

**Comment by antfu** [comment]
I was trying to find a reference from Vue's docs, but it seems they didn't mention this caveat at all. While searching online, I didn't find many articles talking about it either. I guess that would be because Vue compiler compiles it in SFC, making only JSX or manual render function would go into this pitfall. (during the research I found a related article I wrote a while ago https://antfu.me/posts/async-with-composition-api that I almost forgot about :P)

I would feel this caveat should be better document in Vue's doc first, and then in Nuxt we could reference to it and adding some Nuxt-specific tips/utils on top.
